### ISSUE TITLE
Research: Budgeted context selection, grounding, and compression for AI coding agents

### ISSUE THREAD

**Issue by DianaTao** [promptdriven/pdd#1439]
### Summary

AI coding agents are increasingly limited by **context quality**, not only model capability. Broad repository ingestion wastes tokens and can distract the model; too little context causes hallucinated APIs, broken interfaces, and hidden test failures.

This issue tracks a research program to study:

> Under a fixed token budget, how do we select, ground, and compress repository context so the model produces better code?

The central claim to test:

> **The right small context may beat a large messy context.**

Primary evaluation metric:

```text
hidden successes per 10k input tokens
```

---

### Problem statement

In real repositories, a coding agent may need more than a task description:

- target source file
- related tests
- dependency interfaces
- API contracts and schemas
- examples of prior fixes
- documentation and architectural rules
- hidden coupling information

Human developers do not read the whole repo before fixing a bug. They selectively inspect the issue, failing test, target function, related APIs, patterns, docs, and similar past fixes. LLM-based agents need comparable behavior.

LLM context windows are not free. More context can increase cost, latency, irrelevant information, confusion from stale files, and loss of focus. The goal is **correctness per token**, not maximum context.

---

#### A. Context selection

Choose which repository artifacts to include in the prompt under a token budget.

Candidate ranking signals:

- task similarity
- file path relevance
- import graph proximity
- test-to-source mapping
- contract IDs
- example retrieval score
- validation status
- token cost
- staleness risk

Simplified scoring objective:

```text
context_score =
    relevance
  + contract_value
  + verifier_value
  + grounding_value
  - token_cost
  - staleness_risk

Maximize expected verifier success subject to a token budget.
```

#### B. Grounding

Provide concrete evidence so the model does not invent APIs, patterns, or assumptions.

Grounding sources:

- usage examples
- prior validated fixes
- API contracts and type signatures
- schemas
- tests
- docs and architectural rules

#### C. Contract-level compression

Instead of including an entire dependency file, include only the **contract** that matters: public interfaces, schemas, and explicit rules.

Example compression:

```python
# Instead of a 2,000-line payment_client.py, include:
class PaymentClient:
    def create_payment_intent(
        self,
        customer_id: str,
        amount_cents: int,
        currency: str
    ) -> PaymentIntent: ...

    def refund_payment(
        self,
        payment_id: str,
        reason: str | None = None
    ) -> RefundResult: ...
```

With rules:

```text
MUST pass amount in cents.
MUST NOT pass raw dollar float.
MUST use customer_id, not user_id.
Refunds are only valid for succeeded payments.
```

This is **dependency and API slicing**: preserve interface signal, remove implementation noise.

---

### Hypothesis

> AI coding agents can achieve higher correctness per token when context is selected, grounded, and compressed around task intent, tests, public interfaces, and dependency contracts, instead of using broad repo search or full-file inclusion.

---

#### RQ1: Can we automatically select the right context under budget?

- Which context pieces have the highest marginal value per token?
- Does selected context reduce files read, retries, cost, and runtime?
- Does a context manifest make runs easier to review and reproduce?
- How does performance degrade as repositories grow noisier?

#### RQ2: Does grounding reduce hallucination?

- Do examples and contracts improve correct use of APIs, imports, schemas, types, and architectural patterns?
- What is the rate of invented modules, wrong imports, and schema misuse?

#### RQ3: Can compression preserve useful information?

Representations to compare:

- full target file
- target API slice
- dependency contract slice
- selected test assertions
- example slice
- AST/symbol summary
- full dependency source (fallback only when needed)

Hard question:

> What can we remove without hurting hidden-test success?

---

### Relationship to existing PDD capabilities

This research should build on and evaluate existing mechanisms rather than reinvent them:

| Area | Existing PDD surface |
|------|----------------------|
| Structural extraction | `ContentSelector` (`def:`, `class:`, `mode="interface"`) |
| Dependency discovery | `auto_include` with `select=` / `query=` |
| Compression modes | `--context-compression` (`off`, `test`, `examples`, `contracts`, `all`) |
| Contract parsing | `contract_ir`, `contract_check`, prompt `<contract_rules>` |
| Bounded sync context | `compressed_sync_context.py` with token budget |
| Auditability | `evidence_manifest.py`, grounding provenance |

The research contribution is not "add context engineering" in the abstract. It is a **controlled comparison** of selection, grounding, and compression strategies under a fixed budget, with reproducible manifests and a failure taxonomy.

---

#### Baselines

| Arm | Contents | Expected weakness |
|-----|----------|-------------------|
| **B1: Prompt-only** | Issue/task description only | High hallucination risk; low repo awareness |
| **B2: Prompt + target source** | Issue + main file to edit | May miss contracts, tests, hidden coupling |
| **B3: Prompt + source + visible tests** | Common practical baseline | May overfit public tests; still fail hidden tests |
| **B4: Broad repo context** | Many nearby files or agentic file observations | Expensive, noisy, harder to review |

#### Proposed arms

| Arm | Contents |
|-----|----------|
| **P1: Budgeted selector** | Task-aware file/snippet selection under token budget |
| **P2: Budgeted selector + example grounding** | P1 + prior examples / validated fixes |
| **P3: Budgeted selector + contract grounding** | P1 + dependency contracts, public APIs, schemas, rules |
| **P4: Full proposed system** | P1 + examples + dependency slices + context manifest |

P4 is the main proposed system:

```text
task-aware retrieval
+ selected tests
+ examples
+ target API slice
+ dependency contract slice
+ context budget
+ context manifest
```

---

#### A. Correctness

```text
public pass rate
hidden pass rate
public-hidden generalization gap
interface preservation rate
hidden coupling failure rate
hallucinated API/import/schema rate
hidden successes per 10k input tokens   # primary
```

#### B. Context efficiency

```text
input tokens per task
input tokens per hidden success
files included
files read before first edit
context budget utilization
useful context ratio
dropped-context audit
```

#### C. Economic

```text
cost per hidden success
wall-clock seconds per hidden success
retry count per success
candidate count per success
```

#### D. Reviewability

```text
context manifest completeness
scope violation rate
generated diff size
reviewer-relevant artifact count
```

The context manifest should answer:

```text
What did the model see?
Why was it selected?
What was excluded?
How many tokens did it cost?
```

---

### Failure taxonomy

Failures should be classified, not only reported as pass/fail:

| Failure type | Meaning |
|--------------|---------|
| Missing dependency context | Model lacked behavior of a related module |
| Wrong API/schema | Invented or misused an interface |
| Stale grounding | Retrieved example or contract was outdated |
| Weak public tests | Public tests passed; hidden tests failed |
| Over-compression | Slice removed necessary behavior |
| Wrong target localization | Edited wrong file or function |

---

### Example walkthrough

**Task:** Fix refund validation so refunds are blocked after timeout.

**Naive context:**

```text
Issue description
refund_service.py
```

**Likely misses:**

```text
merchant-specific timeout policy
UTC timestamp handling
refund timeout in hours, not days
only SUCCEEDED payments are refundable
distinct error code for failed payments
```

**Better context bundle:**

```text
Task prompt
refund_service.py
test_refund_timeout.py
merchant_policy schema
PaymentIntent status enum
Refund API contract
previous timeout validation example
```

**Compressed contract excerpt:**

```text
Merchant.refund_timeout_hours: int | None
None means no timeout.
PaymentIntent status must be SUCCEEDED before refund.
Use timezone-aware UTC timestamps.
Return REFUND_WINDOW_EXPIRED when timeout has passed.
```

---

### Recommended pilot (30 tasks)

```text
10 backend validation / business-rule tasks
10 API adapter / wrapper tasks
10 historical PR replay tasks
```

Run four arms:

```text
A. Prompt + source + tests          (B3)
B. Broad full-file context          (B4)
C. Budgeted selector only           (P1)
D. Budgeted selector + examples + dependency slices (P4)
```

**Primary metric:** `hidden successes per 10k input tokens`

**Secondary metrics:**

```text
hidden pass rate
input tokens per success
hallucinated API rate
wrong import/schema rate
public-hidden generalization gap
```

**Target claim for pilot:**

> Contract-aware budgeted context selection achieves comparable or better hidden-test success than broad context while using substantially fewer input tokens.

---

### Risks and mitigations

| Risk | Mitigation |
|------|------------|
| Scope too broad | Frame as correctness-per-token comparison across defined arms |
| Weak benchmark | Curate tasks with public/hidden tests, known targets, snapshots, optional contracts, failure labels |
| Hidden tests alone insufficient | Add manual failure analysis, API/schema checks, diff review, typecheck, lint |
| Stale examples hurt | Track example age, validation status, similarity, dependency overlap, test pass status |
| Over-compression drops needed behavior | Contract slice first; escalate to full dependency source when contract is insufficient |

---

### Deliverables

- [ ] 30-task evaluation corpus with public/hidden tests and labeled target files
- [ ] Context manifest schema documenting included/excluded artifacts and selection rationale
- [ ] Benchmark harness running B1–B4 and P1–P4 arms with fixed token budgets
- [ ] Results report: primary metric, secondary metrics, failure taxonomy breakdown
- [ ] Design note in `docs/` summarizing findings and recommended default PDD context policy

---

### Out of scope (for pilot)

- New model training or fine-tuning
- Open-ended agentic repo search without budget controls
- Production deployment changes before pilot results
- Benchmark tasks without hidden tests or failure labels

---

### Acceptance criteria

1. All four pilot arms run on the same 30-task corpus with comparable model/config settings.
2. Every run emits a machine-readable context manifest.
3. Results include `hidden successes per 10k input tokens` as the headline metric.
4. Failures are tagged with the taxonomy above (minimum 80% labeled for pilot report).
5. P4 vs B4 comparison is reported with token cost, hidden pass rate, and hallucination rate.

---

### Key terms

| Term | Meaning |
|------|---------|
| Context selection | Choosing which files/snippets/docs/tests to include |
| Token budget | Maximum allowed input size |
| Grounding | Concrete evidence to prevent hallucination |
| Contract slice | Compressed interface/API/schema representation |
| Dependency slicing | Relevant dependency information only |
| Hidden-test success | Generated code passes unseen tests |
| Public-hidden gap | Visible vs hidden test success difference |
| Context manifest | Record of included/excluded context and reasons |
| Staleness risk | Risk that retrieved examples/docs are outdated |
| Correctness per token | Success rate relative to input token cost |

**Comment by DianaTao** [comment]
Diana June 8-1. take question from https://github.com/promptdriven/pdd_cloud/blob/7443bed89a0ad84ef2b78d40352d637c2ba9c1a9/research/pdd_research_priorities.md
2. Literature review
